File: articles/calculus-for-systems-modeling/articles/systems-of-differential-equations/julia/coupled_system_audit.jl
function predator_prey_rates(prey, predator, alpha, beta, delta, gamma)
    prey_rate = alpha * prey - beta * prey * predator
    predator_rate = delta * prey * predator - gamma * predator
    prey_rate, predator_rate
end

function simulate_predator_prey(prey0, predator0, alpha, beta, delta, gamma, dt, steps)
    prey = prey0
    predator = predator0
    rows = []
    for n in 0:steps
        t = n * dt
        prey_rate, predator_rate = predator_prey_rates(prey, predator, alpha, beta, delta, gamma)
        push!(rows, ("predator_prey_coupled_system", t, prey, predator, prey_rate, predator_rate, alpha, beta, delta, gamma, "explicit_euler", "Predator-prey terms are illustrative and assume continuous well-mixed interaction."))
        prey = max(0.0, prey + dt * prey_rate)
        predator = max(0.0, predator + dt * predator_rate)
    end
    rows
end

println("scenario,time,prey,predator,prey_rate,predator_rate,alpha,beta,delta,gamma,method,warning")
for row in simulate_predator_prey(40.0, 9.0, 0.7, 0.05, 0.02, 0.5, 0.01, 2000)
    println(join(row, ","))
end

File: articles/calculus-for-systems-modeling/articles/systems-of-differential-equations/r/coupled_system_audit.R
predator_prey_rates <- function(prey, predator, alpha, beta, delta, gamma) {
  prey_rate <- alpha * prey - beta * prey * predator
  predator_rate <- delta * prey * predator - gamma * predator
  c(prey_rate = prey_rate, predator_rate = predator_rate)
}

simulate_predator_prey <- function(prey0, predator0, alpha, beta, delta, gamma, dt, steps) {
  prey <- prey0
  predator <- predator0
  rows <- list()

  for (n in 0:steps) {
    t <- n * dt
    rates <- predator_prey_rates(prey, predator, alpha, beta, delta, gamma)

    rows[[length(rows) + 1]] <- data.frame(
      scenario = "predator_prey_coupled_system",
      time = t,
      prey = prey,
      predator = predator,
      prey_rate = rates[["prey_rate"]],
      predator_rate = rates[["predator_rate"]],
      alpha = alpha,
      beta = beta,
      delta = delta,
      gamma = gamma,
      method = "explicit_euler",
      warning = "Predator-prey terms are illustrative and assume continuous well-mixed interaction."
    )

    prey <- max(0, prey + dt * rates[["prey_rate"]])
    predator <- max(0, predator + dt * rates[["predator_rate"]])
  }

  do.call(rbind, rows)
}

results <- simulate_predator_prey(
  prey0 = 40,
  predator0 = 9,
  alpha = 0.7,
  beta = 0.05,
  delta = 0.02,
  gamma = 0.5,
  dt = 0.01,
  steps = 2000
)

dir.create("outputs/tables", recursive = TRUE, showWarnings = FALSE)
write.csv(results, "outputs/tables/r_coupled_system_audit.csv", row.names = FALSE)
print(head(results))
print(tail(results))
